GitHub Actions caching causing slowdown and spurious failures in CI

The CI job steps using GitHub Actions `actions/cache@v1.1.2` are intermittently running very slowly, and occasionally causing spurious failures due to failed cache restoration.

This has been occurring for many weeks now.

A recent example can be found here: https://github.com/lampepfl/dotty/runs/1375585886?check_suite_focus=true

![ci_cache_fail](https://user-images.githubusercontent.com/17014833/98737290-db683300-235a-11eb-80f9-0df21714ef33.png)

In this case, 45 minutes were spent attempting to restore the cache, and to top if off, an error occurred during the `Cache SBT` action (after 31 minutes!):

```
gzip: stdin: unexpected end of file
/usr/bin/tar: Unexpected EOF in archive
/usr/bin/tar: Unexpected EOF in archive
/usr/bin/tar: Error is not recoverable: exiting now
Warning: Tar failed with error: The process '/usr/bin/tar' failed with exit code 2
```

which I believe is the cause of this subsequent failure in the Test step:

```
Exception in thread "sbt-parser-init-thread" java.lang.ExceptionInInitializerError
	at sbt.internal.parser.SbtParserInit$$anon$2.run(SbtParser.scala:171)
Caused by: scala.reflect.internal.FatalError: Error accessing /root/.sbt/0.13/java9-rt-ext-private_build_14_0_1/rt.jar
  ...
Caused by: java.util.zip.ZipException: zip END header not found
  ...
Error:  java.lang.NoClassDefFoundError: Could not initialize class sbt.internal.parser.SbtParser$
Error:  Use 'last' for the full log.
Project loading failed: (r)etry, (q)uit, (l)ast, or (i)gnore?
```

--- a/.github/workflows/ci.yaml
+++ b/.github/workflows/ci.yaml
@@ -17,7 +17,7 @@ jobs:
            startsWith(github.event.ref, 'refs/tags/sbt-dotty-'))"
 
     steps:
-      - name: Set JDK 11 as default
+      - name: Set JDK 14 as default
         run: echo "/usr/lib/jvm/java-14-openjdk-amd64/bin" >> $GITHUB_PATH
 
       - name: Checkout cleanup script
@@ -62,7 +62,7 @@ jobs:
            startsWith(github.event.ref, 'refs/tags/sbt-dotty-'))"
 
     steps:
-      - name: Set JDK 11 as default
+      - name: Set JDK 14 as default
         run: echo "/usr/lib/jvm/java-14-openjdk-amd64/bin" >> $GITHUB_PATH
 
       - name: Checkout cleanup script

--- a/.github/workflows/ci.yaml
+++ b/.github/workflows/ci.yaml
@@ -12,7 +12,12 @@ env:
 jobs:
   test:
     runs-on: [self-hosted, Linux]
-    container: lampepfl/dotty:2020-09-08
+    container:
+      image: lampepfl/dotty:2020-09-08
+      volumes:
+        - ${{ github.workspace }}/../../cache/sbt:/root/.sbt
+        - ${{ github.workspace }}/../../cache/ivy:/root/.ivy2/cache
+        - ${{ github.workspace }}/../../cache/general:/root/.cache
     if: "!(github.event_name == 'push' &&
            startsWith(github.event.ref, 'refs/tags/sbt-dotty-'))"
 
@@ -28,27 +33,6 @@ jobs:
 
       - name: Git Checkout
         uses: actions/checkout@v2
-
-      - name: Cache Ivy
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.ivy2/cache
-          key: ${{ runner.os }}-ivy-${{ hashFiles('**/build.sbt') }}
-          restore-keys: ${{ runner.os }}-ivy-
-
-      - name: Cache SBT
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.sbt
-          key: ${{ runner.os }}-sbt-${{ hashFiles('project/**') }}
-          restore-keys: ${{ runner.os }}-sbt-
-
-      - name: Cache Coursier and Mill
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.cache
-          key: ${{ runner.os }}-general-${{ hashFiles('**/build.sbt') }}
-          restore-keys: ${{ runner.os }}-general-
 
       - name: Test
         run: |
@@ -57,7 +41,12 @@ jobs:
 
   test_bootstrapped:
     runs-on: [self-hosted, Linux]
-    container: lampepfl/dotty:2020-09-08
+    container:
+      image: lampepfl/dotty:2020-09-08
+      volumes:
+        - ${{ github.workspace }}/../../cache/sbt:/root/.sbt
+        - ${{ github.workspace }}/../../cache/ivy:/root/.ivy2/cache
+        - ${{ github.workspace }}/../../cache/general:/root/.cache
     if: "!(github.event_name == 'push' &&
            startsWith(github.event.ref, 'refs/tags/sbt-dotty-'))"
 
@@ -73,27 +62,6 @@ jobs:
 
       - name: Git Checkout
         uses: actions/checkout@v2
-
-      - name: Cache Ivy
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.ivy2/cache
-          key: ${{ runner.os }}-ivy-${{ hashFiles('**/build.sbt') }}
-          restore-keys: ${{ runner.os }}-ivy-
-
-      - name: Cache SBT
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.sbt
-          key: ${{ runner.os }}-sbt-${{ hashFiles('project/**') }}
-          restore-keys: ${{ runner.os }}-sbt-
-
-      - name: Cache Coursier and Mill
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.cache
-          key: ${{ runner.os }}-general-${{ hashFiles('**/build.sbt') }}
-          restore-keys: ${{ runner.os }}-general-
 
       - name: Test
         run: |
@@ -135,7 +103,12 @@ jobs:
 
   community_build:
     runs-on: [self-hosted, Linux]
-    container: lampepfl/dotty:2020-09-08
+    container:
+      image: lampepfl/dotty:2020-09-08
+      volumes:
+        - ${{ github.workspace }}/../../cache/sbt:/root/.sbt
+        - ${{ github.workspace }}/../../cache/ivy:/root/.ivy2/cache
+        - ${{ github.workspace }}/../../cache/general:/root/.cache
 
     steps:
       - name: Checkout cleanup script
@@ -147,27 +120,6 @@ jobs:
       - name: Git Checkout
         uses: actions/checkout@v2
 
-      - name: Cache Ivy
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.ivy2/cache
-          key: ${{ runner.os }}-ivy-${{ hashFiles('**/build.sbt') }}
-          restore-keys: ${{ runner.os }}-ivy-
-
-      - name: Cache SBT
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.sbt
-          key: ${{ runner.os }}-sbt-${{ hashFiles('project/**') }}
-          restore-keys: ${{ runner.os }}-sbt-
-
-      - name: Cache Coursier and Mill
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.cache
-          key: ${{ runner.os }}-general-${{ hashFiles('**/build.sbt') }}
-          restore-keys: ${{ runner.os }}-general-
-
       - name: Test
         run: |
           git submodule sync
@@ -176,7 +128,12 @@ jobs:
 
   test_sbt:
     runs-on: [self-hosted, Linux]
-    container: lampepfl/dotty:2020-09-08
+    container:
+      image: lampepfl/dotty:2020-09-08
+      volumes:
+        - ${{ github.workspace }}/../../cache/sbt:/root/.sbt
+        - ${{ github.workspace }}/../../cache/ivy:/root/.ivy2/cache
+        - ${{ github.workspace }}/../../cache/general:/root/.cache
     if: (
           github.event_name == 'push' &&
           startsWith(github.event.ref, 'refs/tags/')
@@ -193,33 +150,18 @@ jobs:
       - name: Git Checkout
         uses: actions/checkout@v2
 
-      - name: Cache Ivy
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.ivy2/cache
-          key: ${{ runner.os }}-ivy-${{ hashFiles('**/build.sbt') }}
-          restore-keys: ${{ runner.os }}-ivy-
-
-      - name: Cache SBT
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.sbt
-          key: ${{ runner.os }}-sbt-${{ hashFiles('project/**') }}
-          restore-keys: ${{ runner.os }}-sbt-
-
-      - name: Cache Coursier and Mill
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.cache
-          key: ${{ runner.os }}-general-${{ hashFiles('**/build.sbt') }}
-          restore-keys: ${{ runner.os }}-general-
-
       - name: Test
         run: ./project/scripts/sbt sbt-dotty/scripted
 
   test_java8:
     runs-on: [self-hosted, Linux]
-    container: lampepfl/dotty:2020-09-08
+    container:
+      image: lampepfl/dotty:2020-09-08
+      volumes:
+        - ${{ github.workspace }}/../../cache/sbt:/root/.sbt
+        - ${{ github.workspace }}/../../cache/ivy:/root/.ivy2/cache
+        - ${{ github.workspace }}/../../cache/general:/root/.cache
+
     if: "(
           github.event_name == 'push' &&
           startsWith(github.event.ref, 'refs/tags/') &&
@@ -240,33 +182,17 @@ jobs:
       - name: Git Checkout
         uses: actions/checkout@v2
 
-      - name: Cache Ivy
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.ivy2/cache
-          key: ${{ runner.os }}-ivy-${{ hashFiles('**/build.sbt') }}
-          restore-keys: ${{ runner.os }}-ivy-
-
-      - name: Cache SBT
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.sbt
-          key: ${{ runner.os }}-sbt-${{ hashFiles('project/**') }}
-          restore-keys: ${{ runner.os }}-sbt-
-
-      - name: Cache Coursier and Mill
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.cache
-          key: ${{ runner.os }}-general-${{ hashFiles('**/build.sbt') }}
-          restore-keys: ${{ runner.os }}-general-
-
       - name: Test
         run: ./project/scripts/sbt ";compile ;test"
 
   publish_nightly:
     runs-on: [self-hosted, Linux]
-    container: lampepfl/dotty:2020-09-08
+    container:
+      image: lampepfl/dotty:2020-09-08
+      volumes:
+        - ${{ github.workspace }}/../../cache/sbt:/root/.sbt
+        - ${{ github.workspace }}/../../cache/ivy:/root/.ivy2/cache
+        - ${{ github.workspace }}/../../cache/general:/root/.cache
     needs: [test, test_bootstrapped, community_build, test_sbt, test_java8]
     if: github.event_name == 'schedule'
     env:
@@ -286,34 +212,18 @@ jobs:
       - name: Git Checkout
         uses: actions/checkout@v2
 
-      - name: Cache Ivy
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.ivy2/cache
-          key: ${{ runner.os }}-ivy-${{ hashFiles('**/build.sbt') }}
-          restore-keys: ${{ runner.os }}-ivy-
-
-      - name: Cache SBT
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.sbt
-          key: ${{ runner.os }}-sbt-${{ hashFiles('project/**') }}
-          restore-keys: ${{ runner.os }}-sbt-
-
-      - name: Cache Coursier and Mill
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.cache
-          key: ${{ runner.os }}-general-${{ hashFiles('**/build.sbt') }}
-          restore-keys: ${{ runner.os }}-general-
-
       - name: Publish Nightly
         run: |
           ./project/scripts/sbtPublish ";project scala3-bootstrapped ;publishSigned ;sonatypeBundleRelease"
 
   nightly_documentation:
     runs-on: [self-hosted, Linux]
-    container: lampepfl/dotty:2020-09-08
+    container:
+      image: lampepfl/dotty:2020-09-08
+      volumes:
+        - ${{ github.workspace }}/../../cache/sbt:/root/.sbt
+        - ${{ github.workspace }}/../../cache/ivy:/root/.ivy2/cache
+        - ${{ github.workspace }}/../../cache/general:/root/.cache
     needs: [publish_nightly]
     if: github.event_name == 'schedule'
     env:
@@ -332,27 +242,6 @@ jobs:
       - name: Git Checkout
         uses: actions/checkout@v2
 
-      - name: Cache Ivy
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.ivy2/cache
-          key: ${{ runner.os }}-ivy-${{ hashFiles('**/build.sbt') }}
-          restore-keys: ${{ runner.os }}-ivy-
-
-      - name: Cache SBT
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.sbt
-          key: ${{ runner.os }}-sbt-${{ hashFiles('project/**') }}
-          restore-keys: ${{ runner.os }}-sbt-
-
-      - name: Cache Coursier and Mill
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.cache
-          key: ${{ runner.os }}-general-${{ hashFiles('**/build.sbt') }}
-          restore-keys: ${{ runner.os }}-general-
-
       - name: Generate Website
         run: |
           ./project/scripts/genDocs -doc-snapshot
@@ -367,7 +256,12 @@ jobs:
 
   publish_release:
     runs-on: [self-hosted, Linux]
-    container: lampepfl/dotty:2020-09-08
+    container:
+      image: lampepfl/dotty:2020-09-08
+      volumes:
+        - ${{ github.workspace }}/../../cache/sbt:/root/.sbt
+        - ${{ github.workspace }}/../../cache/ivy:/root/.ivy2/cache
+        - ${{ github.workspace }}/../../cache/general:/root/.cache
     needs: [test, test_bootstrapped, community_build, test_sbt, test_java8]
     if: github.event_name == 'push' &&
         startsWith(github.event.ref, 'refs/tags/') &&
@@ -389,27 +283,6 @@ jobs:
 
       - name: Git Checkout
         uses: actions/checkout@v2
-
-      - name: Cache Ivy
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.ivy2/cache
-          key: ${{ runner.os }}-ivy-${{ hashFiles('**/build.sbt') }}
-          restore-keys: ${{ runner.os }}-ivy-
-
-      - name: Cache SBT
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.sbt
-          key: ${{ runner.os }}-sbt-${{ hashFiles('project/**') }}
-          restore-keys: ${{ runner.os }}-sbt-
-
-      - name: Cache Coursier and Mill
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.cache
-          key: ${{ runner.os }}-general-${{ hashFiles('**/build.sbt') }}
-          restore-keys: ${{ runner.os }}-general-
 
       - name: Publish Release
         run: |
@@ -462,7 +335,12 @@ jobs:
 
   release_documentation:
     runs-on: [self-hosted, Linux]
-    container: lampepfl/dotty:2020-09-08
+    container:
+      image: lampepfl/dotty:2020-09-08
+      volumes:
+        - ${{ github.workspace }}/../../cache/sbt:/root/.sbt
+        - ${{ github.workspace }}/../../cache/ivy:/root/.ivy2/cache
+        - ${{ github.workspace }}/../../cache/general:/root/.cache
     needs: [publish_release]
     if: github.event_name == 'push' &&
         startsWith(github.event.ref, 'refs/tags/') &&
@@ -484,27 +362,6 @@ jobs:
       - name: Git Checkout
         uses: actions/checkout@v2
 
-      - name: Cache Ivy
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.ivy2/cache
-          key: ${{ runner.os }}-ivy-${{ hashFiles('**/build.sbt') }}
-          restore-keys: ${{ runner.os }}-ivy-
-
-      - name: Cache SBT
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.sbt
-          key: ${{ runner.os }}-sbt-${{ hashFiles('project/**') }}
-          restore-keys: ${{ runner.os }}-sbt-
-
-      - name: Cache Coursier and Mill
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.cache
-          key: ${{ runner.os }}-general-${{ hashFiles('**/build.sbt') }}
-          restore-keys: ${{ runner.os }}-general-
-
       - name: Generate Website
         run: |
           ./project/scripts/genDocs -doc-snapshot
@@ -519,7 +376,12 @@ jobs:
 
   publish_sbt_release:
     runs-on: [self-hosted, Linux]
-    container: lampepfl/dotty:2020-09-08
+    container:
+      image: lampepfl/dotty:2020-09-08
+      volumes:
+        - ${{ github.workspace }}/../../cache/sbt:/root/.sbt
+        - ${{ github.workspace }}/../../cache/ivy:/root/.ivy2/cache
+        - ${{ github.workspace }}/../../cache/general:/root/.cache
     needs: [community_build, test_sbt]
     if: github.event_name == 'push' &&
         startsWith(github.event.ref, 'refs/tags/sbt-dotty-')
@@ -540,27 +402,6 @@ jobs:
 
       - name: Git Checkout
         uses: actions/checkout@v2
-
-      - name: Cache Ivy
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.ivy2/cache
-          key: ${{ runner.os }}-ivy-${{ hashFiles('**/build.sbt') }}
-          restore-keys: ${{ runner.os }}-ivy-
-
-      - name: Cache SBT
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.sbt
-          key: ${{ runner.os }}-sbt-${{ hashFiles('project/**') }}
-          restore-keys: ${{ runner.os }}-sbt-
-
-      - name: Cache Coursier and Mill
-        uses: actions/cache@v1.1.2
-        with:
-          path: /root/.cache
-          key: ${{ runner.os }}-general-${{ hashFiles('**/build.sbt') }}
-          restore-keys: ${{ runner.os }}-general-
 
       - name: Publish Dotty SBT Plugin Release
         run: |

--- a/.github/workflows/ci.yaml
+++ b/.github/workflows/ci.yaml
@@ -34,6 +34,9 @@ jobs:
       - name: Git Checkout
         uses: actions/checkout@v2
 
+      - name: Add SBT proxy repositories
+        run: cp -vf .github/workflows/repositories /root/.sbt/ ; true
+
       - name: Test
         run: |
           ./project/scripts/sbt ";compile ;test"
@@ -62,6 +65,9 @@ jobs:
 
       - name: Git Checkout
         uses: actions/checkout@v2
+
+      - name: Add SBT proxy repositories
+        run: cp -vf .github/workflows/repositories /root/.sbt/ ; true
 
       - name: Test
         run: |
@@ -120,6 +126,9 @@ jobs:
       - name: Git Checkout
         uses: actions/checkout@v2
 
+      - name: Add SBT proxy repositories
+        run: cp -vf .github/workflows/repositories /root/.sbt/ ; true
+
       - name: Test
         run: |
           git submodule sync
@@ -149,6 +158,9 @@ jobs:
 
       - name: Git Checkout
         uses: actions/checkout@v2
+
+      - name: Add SBT proxy repositories
+        run: cp -vf .github/workflows/repositories /root/.sbt/ ; true
 
       - name: Test
         run: ./project/scripts/sbt sbt-dotty/scripted
@@ -182,6 +194,9 @@ jobs:
       - name: Git Checkout
         uses: actions/checkout@v2
 
+      - name: Add SBT proxy repositories
+        run: cp -vf .github/workflows/repositories /root/.sbt/ ; true
+
       - name: Test
         run: ./project/scripts/sbt ";compile ;test"
 
@@ -212,6 +227,9 @@ jobs:
       - name: Git Checkout
         uses: actions/checkout@v2
 
+      - name: Add SBT proxy repositories
+        run: cp -vf .github/workflows/repositories /root/.sbt/ ; true
+
       - name: Publish Nightly
         run: |
           ./project/scripts/sbtPublish ";project scala3-bootstrapped ;publishSigned ;sonatypeBundleRelease"
@@ -241,6 +259,9 @@ jobs:
 
       - name: Git Checkout
         uses: actions/checkout@v2
+
+      - name: Add SBT proxy repositories
+        run: cp -vf .github/workflows/repositories /root/.sbt/ ; true
 
       - name: Generate Website
         run: |
@@ -283,6 +304,9 @@ jobs:
 
       - name: Git Checkout
         uses: actions/checkout@v2
+
+      - name: Add SBT proxy repositories
+        run: cp -vf .github/workflows/repositories /root/.sbt/ ; true
 
       - name: Publish Release
         run: |
@@ -362,6 +386,9 @@ jobs:
       - name: Git Checkout
         uses: actions/checkout@v2
 
+      - name: Add SBT proxy repositories
+        run: cp -vf .github/workflows/repositories /root/.sbt/ ; true
+
       - name: Generate Website
         run: |
           ./project/scripts/genDocs -doc-snapshot
@@ -402,6 +429,9 @@ jobs:
 
       - name: Git Checkout
         uses: actions/checkout@v2
+
+      - name: Add SBT proxy repositories
+        run: cp -vf .github/workflows/repositories /root/.sbt/ ; true
 
       - name: Publish Dotty SBT Plugin Release
         run: |

--- a/.github/workflows/repositories
+++ b/.github/workflows/repositories
@@ -1,0 +1,4 @@
+[repositories]
+local
+my-ivy-proxy-releases: https://scala-webapps.epfl.ch/artifactory/sbt-plugin-releases/, [organization]/[module]/(scala_[scalaVersion]/)(sbt_[sbtVersion]/)[revision]/[type]s/[artifact](-[classifier]).[ext]
+my-maven-proxy-releases: https://scala-webapps.epfl.ch/artifactory/central/

